### PR TITLE
Initial stab at importing idf components

### PR DESCRIPTION
**Pull request by jesserockz**
# What does this implement/fix?

This allows idf-based esp32 builds to reference idf components.

Todo: Allow adding arbitrary components added via config entries, maybe:
```yaml
esp32:
  framework:
    type: esp-idf
    components:
      - name: esp32-camera:
        source: https://github.com/espressif/esp32-camera
        refresh: 1day
```

Thoughts/suggestions?

Here is an example change for `esp32_camera`:
```diff
diff --git a/esphome/components/esp32_camera/__init__.py b/esphome/components/esp32_camera/__init__.py
index 753b6ed9d..786884bc6 100644
--- a/esphome/components/esp32_camera/__init__.py
+++ b/esphome/components/esp32_camera/__init__.py
@@ -16,7 +16,7 @@ from esphome.const import (
     CONF_TRIGGER_ID,
 )
 from esphome.core import CORE
-from esphome.components.esp32 import add_idf_sdkconfig_option
+from esphome.components.esp32 import add_idf_sdkconfig_option, add_idf_component
 from esphome.cpp_helpers import setup_entity
 
 DEPENDENCIES = ["esp32"]
@@ -263,8 +263,11 @@ async def to_code(config):
     cg.add_define("USE_ESP32_CAMERA")
 
     if CORE.using_esp_idf:
-        cg.add_library("espressif/esp32-camera", "1.0.0")
+        add_idf_component(
+            "esp32-camera", "https://github.com/espressif/esp32-camera.git", "v2.0.2"
+        )
         add_idf_sdkconfig_option("CONFIG_RTCIO_SUPPORT_RTC_GPIO_DESC", True)
+        add_idf_sdkconfig_option("CONFIG_SCCB_CLK_FREQ", 100000)
 
     for conf in config.get(CONF_ON_STREAM_START, []):
         trigger = cg.new_Pvariable(conf[CONF_TRIGGER_ID], var)
```

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by spali** [comment]
Love the idea, currently have to import components by cloning the repo and add it as local library via `libraries`.

@jesserockz  
Would this also support "collection" repo of components like https://github.com/espressif/idf-extra-components?

**Comment by jesserockz** [comment]
@spali No, but this is why PR feedback is important. I did not think about this case and when I come back to this PR I will try make sure it works with repos like that.
Thanks
Jesse

**Comment by kbx81** [comment]
Bump 😄

**Comment by kbx81** [comment]
Cool, I just gave this a shot and it appears to work for `mdns` -- I used this in my yaml:

```yaml
esp32:
  board: featheresp32
  framework:
    type: esp-idf
    components:
      - name: mdns
        source: github://espressif/esp-protocols
```

**Review by kbx81 — APPROVED** [review]
I'm not really qualified to do a proper review of Python code, but it works so I approve! 😄

File: esphome/components/esp32/__init__.py [hunk 1]
```diff
@@ -372,6 +414,20 @@ async def to_code(config):
...
+                add_idf_component(
+                    name=component[CONF_NAME],
+                    repo=source[CONF_URL],
+                    ref=source.get(CONF_REF),
+                    path=component.get(CONF_PATH),
+                    refresh=component[CONF_REFRESH],
+                )
+            elif source[CONF_TYPE] == TYPE_LOCAL:
+                _LOGGER.warn("Local components are not implemented yet.")
+                pass
```
Comment (jesserockz) [inline esphome/components/esp32/__init__.py]:
```suggestion
                _LOGGER.warning("Local components are not implemented yet.")
```

**Pull request merged**